packages/instantsearch.js/src/widgets/index/index.ts

```diff
@@ -458,16 +458,31 @@ const index = (widgetParams: IndexWidgetParams): IndexWidget => {
       );
 
       if (localInstantSearchInstance && Boolean(widgets.length)) {
-        const cleanedState = widgets.reduce((state, widget) => {
-          // the `dispose` method exists at this point we already assert it
-          const next = widget.dispose!({
-            helper: helper!,
-            state,
-            parent: this,
-          });
+        const { cleanedSearchState, cleanedRecommendState } = widgets.reduce(
+          (states, widget) => {
+            // the `dispose` method exists at this point we already assert it
+            const next = widget.dispose!({
+              helper: helper!,
+              state: states.cleanedSearchState,
+              recommendState: states.cleanedRecommendState,
+              parent: this,
+            });
 
-          return next || state;
-        }, helper!.state);
+            return {
+              ...states,
+              ...(next instanceof algoliasearchHelper.SearchParameters
```

Comment:
mutating rather than spreading will be quite a bit faster here

```
if (next instanceof algoliasearchHelper.RecommendParameters) {
  states.cleanedRecommendState = next
} else {
  states.cleanedSearchState = next
}
return states
```